clouddriver-kubernetes-v2/src/test/java/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2CredentialsTest.java

```diff
@@ -0,0 +1,376 @@
+    assertThat(timers).hasSize(1);
+
+    Timer timer = timers.get(0);
+    assertThat(timer.id().name()).isEqualTo("kubernetes.api");
+    assertThat(timer.totalTime()).isEqualTo(500);
+  }
+
+  @Test
+  void metricTagsForListThrowingOtherException() {
+    KubectlJobExecutor jobExecutor = mock(KubectlJobExecutor.class);
+    Registry registry = new DefaultRegistry();
+    KubernetesV2Credentials credentials = getCredentials(registry, jobExecutor);
+
+    when(jobExecutor.list(eq(credentials), any(), any(), any()))
+        .thenThrow(new CustomException("Kubernetes error"));
+
+    assertThatThrownBy(
+            () ->
+                credentials.list(
+                    ImmutableList.of(KubernetesKind.DEPLOYMENT, KubernetesKind.REPLICA_SET),
+                    NAMESPACE))
+        .isInstanceOf(KubectlException.class);
+    ImmutableList<Timer> timers = registry.timers().collect(toImmutableList());
+    assertThat(timers).hasSize(1);
+
+    Timer timer = timers.get(0);
+    assertThat(timer.id().name()).isEqualTo("kubernetes.api");
+
+    assertThat(timer.id().tags())
+        .containsExactlyInAnyOrder(
+            Tag.of("account", ACCOUNT_NAME),
+            Tag.of("action", "list"),
+            Tag.of("kinds", "deployment,replicaSet"),
+            Tag.of("namespace", NAMESPACE),
+            Tag.of("success", "false"),
+            Tag.of("reason", "CustomException: Kubernetes error"));
+  }
+
+  @Test
+  void timeRecordedForListThrowingOtherException() {
+    KubectlJobExecutor jobExecutor = mock(KubectlJobExecutor.class);
+
+    ManualClock clock = new ManualClock();
+    Registry registry = new DefaultRegistry(clock);
+    KubernetesV2Credentials credentials = getCredentials(registry, jobExecutor);
+
+    clock.setMonotonicTime(1000);
+    when(jobExecutor.list(eq(credentials), any(), any(), any()))
+        .then(
+            call -> {
+              clock.setMonotonicTime(1500);
+              throw new CustomException("Kubernetes errror");
+            });
+    assertThatThrownBy(
+            () ->
+                credentials.list(
+                    ImmutableList.of(KubernetesKind.DEPLOYMENT, KubernetesKind.REPLICA_SET),
+                    NAMESPACE))
+        .isInstanceOf(KubectlException.class);
+
+    ImmutableList<Timer> timers = registry.timers().collect(toImmutableList());
+    assertThat(timers).hasSize(1);
+
+    Timer timer = timers.get(0);
+    assertThat(timer.id().name()).isEqualTo("kubernetes.api");
+    assertThat(timer.totalTime()).isEqualTo(500);
+  }
+
+  @Test
+  void propagatedExceptionForListThrowingOtherException() {
+    KubectlJobExecutor jobExecutor = mock(KubectlJobExecutor.class);
+    Registry registry = new DefaultRegistry();
+    KubernetesV2Credentials credentials = getCredentials(registry, jobExecutor);
+
+    Exception cause = new CustomException("Kubernetes error");
+    when(jobExecutor.list(eq(credentials), any(), any(), any())).thenThrow(cause);
```

Comment:
Probably doesn't matter at all, but is there any reason we have `any()` for the kinds and namespace parameters in these tests when we're already passing them to the credentials call below?

Reply:
My intent here was to avoid over-specifying the tests and focus on having each test verify the behavior it's testing.  In this case the test is ensuring that an exception thrown by `jobExecutor` gets propagated properly out of the `credentials.list` call, so it seemed best to just have `jobExecutor.list` always throw that exception rather than have it coupled to exactly the arguments that are passed.  (Then, for example, if we changed how `credentials.list` passed things to `jobExecutor.list` the test on exception propagation would break even though exception propagation hasn't actually broken.)

Ideally there would *also* be a test that `credentials.list` properly forwards calls to `kubectl`...in that case we probably would want to assert that the right arguments are forwarded onwards to `jobExecutor.list`, to make that test break if we stopped passing the right arguments.  So you're entirely correct, that these tests don't in any way test that `credentials.list` or `credentials.deploy` do the right thing, but I think that would be belong in other tests (that one of us will probably write one day 😀 )

Reply:
💯 Makes sense; the rule I am deriving from this is to use `any()` instead of explicit argument matchers unless 1) the goal of the test is to assert that the method was called with particular arguments, or 2) multiple `when`s of the same method need to be disambiguated by argument